File: DynamicMapping/lidar.py
import numpy as np
from pose_2d import Pose2D
# from robot import Robot
# from scene import Scene

class Lidar:
    def __init__(self, 
                 id : int,
                 name : str,
                 local_pose : Pose2D, 
                 min_range : float,
                 max_range : float,
                 radius : float = 0.075,
                 color : tuple = (0, 0, 0)):
        
        """
        Instantiates a new LIDAR

            :param id: Unique id of the LIDAR
            :param name: Name of the LIDAR
            :param local_pose: Pose of the LIDAR on the robot
            :param min_range: Min range of the LIDAR (in meters)
            :param max_range: Max range of the LIDAR (in meters)
            :param radius: Radius of the LIDAR in meters (for drawing)
            :param color: Color of the LIDAR (for drawing)
        """
        
        self.id = id
        self.name = name
        self.local_pose = local_pose
        self.global_pose = None
        self.min_range = min_range
        self.max_range = max_range
        self.radius = radius
        self.color = color
        
        # LIDAR observations are represented using a dict of ndarrays
        # for performance reasons:
        self.observations = None

    def update(self, robot_pose : Pose2D, observations : dict):
        """ Updates the global pose and the observations of the LIDAR """

        self.global_pose = self.local_pose.get_global_pose(robot_pose)
        self.observations = observations

    def draw(self, scene : 'Scene', draw_rays : bool):
        """ Draws a circle representing the LIDAR and optionnaly the LIDAR rays """
        
        # Draw the LIDAR rays:
        if draw_rays and self.observations is not None:
            rays_color = (255, 0, 0)

            for range, angle in zip(self.observations['ranges'], self.observations['angles']):
                # Get the position of the observation:
                observation_x, observation_y = self.observation_position(range, angle)

                # DRaw a line between the LIDAR and the observation position:
                scene.draw_line(self.global_pose.x, self.global_pose.y, 
                                observation_x, observation_y, rays_color)

        # Draw the LIDAR itself:
        scene.draw_circle(self.global_pose.x, self.global_pose.y, self.radius, self.color)

    def observation_position(self, range, angle):
        """
        Computes the global position of the given observation made by the LIDAR
        
            :param range: Range of the observation in meters
            :param angle: Angle between the LIDAR forward and the observation (counterclockwise),
                in radians
            :returns: The position of the observation as a tuple (x, y)
        """

        # Compute the global angle of the observation:
        theta = self.global_pose.angle + angle

        x = self.global_pose.x + range * np.cos(theta)
        y = self.global_pose.y + range * np.sin(theta)

        return x, y

    

File: IDP - Dynamic Mapping/Assets/Scripts/Robot/Lidar/Lidar.cs
using System.Collections.Generic;
using UnityEngine;

/// <summary>
/// Class used to represent a LIDAR using Unity raycasting
/// </summary>

public class Lidar {

    // GameObject representing the LIDAR:
    private GameObject lidar;

    // Index of this LIDAR (since we may want to use multiple LIDARs):
    private int lidarIndex;

    // Number of raycasts produced by the LIDAR:
    private int raycastCount;

    // Minimum valid range of the LIDAR (smaller ranges should be discarded):
    private float minRange;

    // Maximum valid range of the LIDAR (bigger ranges should be clamped):
    private float maxRange;

    // List of observations made by the LIDAR:
    private List<Observation> observations;

    // Used for drawing only:
    private Vector3 lastScanPosition, lastScanForward;  // Real position of the LIDAR during last scan

    public Lidar(GameObject lidar, int raycastCount, float minRange, float maxRange, int lidarIndex) {
        this.lidar = lidar;
        this.lidarIndex = lidarIndex;
        this.raycastCount = raycastCount;
        this.minRange = minRange;
        this.maxRange = maxRange;
    }

    /// <summary>
    /// Draws the rays of the LIDAR if drawRays is true
    /// </summary>
    public void DrawGizmos(bool drawRays) {
        if (drawRays && observations != null) {
            Vector3 direction = lastScanForward;

            foreach (Observation observation in observations) {
                Gizmos.color = Color.red;
                Gizmos.DrawRay(lastScanPosition, direction * observation.r);

                direction = Quaternion.AngleAxis(-360f / raycastCount, Vector3.up) * direction;
            }
        }
    }

    /// <summary>
    /// Use raycasting to compute the observations made by the LIDAR. Observations with a range
    /// smaller than minRange are ignored and observations with a range larger than maxRange are
    /// clamped and marked as outOfRange
    /// </summary>
    public Observation[] ComputeObservations() {
        observations = new List<Observation>(raycastCount);

        // Save the current position and orientation of the LIDAR, to draw gizmos later:
        lastScanPosition = lidar.transform.position;
        lastScanForward = lidar.transform.TransformDirection(Vector3.forward);

        // Use raycasting to compute the current observations:
        float observationAngle = 0;
        Vector3 direction = lidar.transform.TransformDirection(Vector3.forward);

        for (int i = 0; i < raycastCount; i++) {
            RaycastHit hit;
            if (Physics.Raycast(lidar.transform.position, direction, out hit, maxRange + 1)
                && hit.distance <= maxRange) {

                // The observation is added only if the range is bigger than minRange:
                if(hit.distance >= minRange)
                    observations.Add(new Observation(hit.distance, observationAngle, lidarIndex, false));
            }
            else {
                // If there was no hit, the range is clamped at maxRange, and the observation is
                // marked as out of range:
                observations.Add(new Observation(maxRange, observationAngle, lidarIndex, true));
            }

            // Rotate the direction of the raycast counterclockwise:
            direction = Quaternion.AngleAxis(-360f / raycastCount, Vector3.up) * direction;
            observationAngle += 2 * Mathf.PI / raycastCount;
        }

        return observations.ToArray();
    }

    /// <summary>
    /// Returns the local pose of the LIDAR on the robot
    /// </summary>
    public Pose2D GetLocalPose() {
        float x = lidar.transform.localPosition.z;
        float y = -lidar.transform.localPosition.x;
        float angle = -lidar.transform.localRotation.eulerAngles.y * Mathf.Deg2Rad;

        return new Pose2D(x, y, angle);
    }

    /// <summary>
    /// Returns the setup of the robot (pose, name, min range, max range)
    /// </summary>
    public LidarSetup GetSetup() {
        return new LidarSetup(lidarIndex, lidar.name, GetLocalPose(), minRange, maxRange);
    }

    /// <summary>
    /// Draws an observation with the given color and for the given duration (used for debugging)
    /// </summary>
    /// <param name="observation">The observation to draw</param>
    /// <param name="color">The color of the duratio</param>
    /// <param name="duration">How long the observation will be visible for (in seconds)</param>
    public void DrawObservation(Observation observation, Color color, float duration) {
        Vector3 direction = lidar.transform.TransformDirection(Vector3.forward);
        direction = Quaternion.AngleAxis(-observation.theta * Mathf.Rad2Deg, Vector3.up) * direction;

        Debug.DrawRay(lidar.transform.position, direction * observation.r, color, duration);
    }    
}
